pkg/genericcli/context.go

```diff
@@ -0,0 +1,840 @@
+	Timeout        *time.Duration `json:"timeout,omitempty" yaml:"timeout,omitempty"`
+	Provider       string         `json:"provider" yaml:"provider"`
+	IsCurrent      bool           `json:"-" yaml:"-"`
+}
+
+type ContextConfig struct {
+	BinaryName    string
+	ConfigName    string
+	ConfigDirName string
+	Fs            afero.Fs
+
+	// I/O
+	DescribePrinter       func() printers.Printer
+	ListPrinter           func() printers.Printer
+	In                    io.Reader
+	Out                   io.Writer
+	ProjectListCompletion cobra.CompletionFunc
+}
+
+type ContextManager struct {
+	cfg *ContextConfig
+}
+
+type ContextUpdateRequest struct {
+	// Name is the ID
+	Name string
+
+	// Fields to patch
+	APIURL         *string
+	APIToken       *string // Pointer, even though Context.APIToken is string
+	DefaultProject *string // Pointer, even though Context.DefaultProject is string
+	Timeout        *time.Duration
+	Provider       *string // Pointer, even though Context.Provider is string
+
+	// Meta-flags for the operation
+	Activate bool
+}
+
+// getFromViper is a helper function to set ContextUpdateRequest fields
+func getFromViper[T any](key string, getFunc func(string) T) *T {
+	if viper.IsSet(key) {
+		return pointer.Pointer(getFunc(key))
+	}
+	return nil
+}
+
+// successCheck returns a green checkmark string.
+func successCheck() string {
+	return color.GreenString("✔")
+}
+
+// NewContextCmd creates the context command tree using genericcli
+func NewContextCmd(c *ContextConfig) *cobra.Command {
+	wrapper := NewContextManager(c)
+
+	cmd := NewCmds(&CmdsConfig[
+		*Context,
+		*ContextUpdateRequest,
+		*Context,
+	]{
+		GenericCLI:      NewGenericCLI(wrapper),
+		BinaryName:      c.BinaryName,
+		Singular:        "context",
+		Plural:          "contexts",
+		Description:     "Manage CLI contexts. A context defines the connection properties (API URL, token, etc.) for a backend. Use \"-\" to switch to the previously used context.",
+		Aliases:         []string{"ctx"},
+		Args:            []string{keyName},
+		Sorter:          contextSorter(),
+		DescribePrinter: c.DescribePrinter,
+		ListPrinter:     c.ListPrinter,
+		In:              c.In,
+		Out:             c.Out,
+		OnlyCmds: OnlyCmds(
+			DescribeCmd,
+			ListCmd,
+			CreateCmd,
+			UpdateCmd,
+			DeleteCmd,
+		),
+		RootCmdMutateFn: func(cmd *cobra.Command) {
+			cmd.Args = cobra.MaximumNArgs(1)
+			cmd.RunE = func(cmd *cobra.Command, args []string) error {
+				// '$ BinaryName context' (no args) should be equal to '$ BinaryName context list'
+				if len(args) == 0 {
+					listCmd, _, err := cmd.Find(pointer.WrapInSlice(string(ListCmd)))
+					if err != nil {
+						return fmt.Errorf(errMsgListCommandNotFound, err)
+					}
+					return listCmd.RunE(listCmd, []string{})
+				}
+
+				// '$ BinaryName context -' or '$ BinaryName context <name>' should behave like 'switch'
+				// Now we can only have one arg thanks to cobra.MaximumNArgs(1) above
+				return wrapper.switchContext(args)
```

Comment:
For later: maybe it's possible to implement this differently. Feels a bit hacky.